File: 2021/java/utils/Day.java
/**
 * Class for preparing Scanner which will read data from text file.
 */

package utils;

import java.io.File;
import java.io.IOException;
import java.util.Scanner;

public abstract class Day {
    protected Scanner scanner;

    protected Day() {
        this.scanner = this.getScanner();
    }

    private Scanner getScanner() {
        Scanner scanner = null;
        String classPath;
        String fileSeparator;
        String dataDir;
        String directoryPath;
        String className;
        String fileName;

        // Get directory path & data filename
        classPath = System.getProperty("java.class.path");
        fileSeparator = File.separator;
        dataDir = "data".concat(fileSeparator);
        directoryPath = classPath.split("java")[0].concat(dataDir);
        className = this.getClass().getSimpleName().toLowerCase();
        fileName = className.concat(".txt");

        // Open input data from text file into Scanner
        try {
            scanner = new Scanner(new File(directoryPath.concat(fileName)));
        } catch (IOException error) {
            System.out.println("Unable to open file.");
            error.printStackTrace();
            System.exit(0);
        }

        return scanner;
    }

}

File: 2021/python/day.py
"""
Class for reading data from text file.
"""
from pathlib import Path
import os


class Day(object):
    def __init__(self):
        self._directory_path = os.path.join(Path(__file__).parents[1].resolve(), "data")
        self._data_filename = f"{self.__class__.__name__.lower()}.txt"
        self._data = self._read_data()

    def _read_data(self) -> list[str]:
        # Read input data from text file
        with open(os.path.join(self._directory_path, self._data_filename), "r") as file:
            return file.read().splitlines()
